### PR TITLE
Add OWASP suppression date recipes

### PR DESCRIPTION
**Pull request by kmccarp**

### CODE REVIEW

File: rewrite-xml/src/main/java/org/openrewrite/xml/security/DateBoundSuppressions.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        return "Adds an expiration date to all OWASP suppressions in order to ensure that they are periodically reviewed. " +
+                "More details: https://jeremylong.github.io/DependencyCheck/general/suppression.html";
+    }
+
+    @Option(displayName = "Until date",
+            description = "Optional. The date to add to the suppression. Default will be 30 days from today.",
+            example = "2023-01-01")
+    @Nullable
+    String untilDate;
+
```
Comment (sambsnyd) [inline rewrite-xml/src/main/java/org/openrewrite/xml/security/DateBoundSuppressions.java:55]: Do these suppression files have a particular name or location within a repository?

Feels like there should be an applicability test preventing this from modifying/wasting time traversing totally unrelated XML files that happen to have a tag named "suppress".

File: rewrite-xml/src/main/java/org/openrewrite/xml/security/DateBoundSuppressions.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+                for (Xml.Attribute attribute : attributes) {
+                    if (attribute.getKeyAsString().equals("until")) {
+                        hasUntil = true;
+                    }
+                    newAttributes.add(attribute);
+                }
+                if (!hasUntil) {
+                    String date = untilDate != null ? untilDate : LocalDate.now().plus(30, ChronoUnit.DAYS).toString();
+                    newAttributes.add(new Xml.Attribute(UUID.randomUUID(), "", Markers.EMPTY,
+                            new Xml.Ident(UUID.randomUUID(), " ", Markers.EMPTY, "until"),
```
Comment (sambsnyd) [inline rewrite-xml/src/main/java/org/openrewrite/xml/security/DateBoundSuppressions.java:78]: Prefer `Tree.randomId()` to `UUID.randomUUID()` as that allows us to easily change the algorithm to generate the GUIDs if we happen to find one more suitable for our purposes.

File: rewrite-xml/src/test/java/org/openrewrite/xml/security/DateBoundSuppressionsTest.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+                spec -> spec.recipe(new DateBoundSuppressions(null)),
+                xml("""
+                        <?xml version="1.0" encoding="UTF-8" ?>
+                        <suppressions>
+                            <suppress>
+                                <notes>
+                                </notes>
+                            </suppress>
+                        </suppressions>""", "" +
+                        "<?xml version=\"1.0\" encoding=\"UTF-8\" ?>\n" +
```
Comment (sambsnyd) [inline rewrite-xml/src/test/java/org/openrewrite/xml/security/DateBoundSuppressionsTest.java:64]: Multi-line """ String for before and concatenated single strings for the after? Not a big deal, but does look a smidge strange

**Review by sambsnyd — CHANGES_REQUESTED** [review]
Nice work

File: rewrite-xml/src/main/java/org/openrewrite/xml/security/DateBoundSuppressions.java (hunk 1 above)
Comment (kmccarp) [reply rewrite-xml/src/main/java/org/openrewrite/xml/security/DateBoundSuppressions.java:55]: Good catch, has to be at the root of the repo, and has to have the xsd `https://jeremylong.github.io/DependencyCheck/dependency-suppression.1.3.xsd` (although i'm thinking to regex match so it isn't required to have 1.3). Will add an applicability test (is there a `suppressions.xml` at the root of the repo?) and a single source applicability test (is this the `suppressions.xml` and does it have `xmlns=https://jeremylong.github.io/DependencyCheck/dependency-suppression.(.*?).xsd`?).

File: rewrite-xml/src/main/java/org/openrewrite/xml/security/DateBoundSuppressions.java (hunk 2 above)
Comment (kmccarp) [reply rewrite-xml/src/main/java/org/openrewrite/xml/security/DateBoundSuppressions.java:78]: Thanks! Also I'm apprehensive about this in general, should I be taking these `prefixUnsafe` or `Quote` from somewhere else? A style perhaps?

File: rewrite-xml/src/test/java/org/openrewrite/xml/security/DateBoundSuppressionsTest.java (hunk 3 above)
Comment (kmccarp) [reply rewrite-xml/src/test/java/org/openrewrite/xml/security/DateBoundSuppressionsTest.java:64]: I did multi-line strings where I could; i.e. where I wasn't doing string concatenation. IntelliJ pings you to use them where you can.

File: rewrite-xml/src/main/java/org/openrewrite/xml/security/DateBoundSuppressions.java (hunk 2 above)
Comment (sambsnyd) [reply rewrite-xml/src/main/java/org/openrewrite/xml/security/DateBoundSuppressions.java:78]: Use `XmlVisitor.autoFormat()`, it will do the necessary style lookup and application for you

File: rewrite-xml/src/test/java/org/openrewrite/xml/security/DateBoundSuppressionsTest.java (hunk 3 above)
Comment (sambsnyd) [reply rewrite-xml/src/test/java/org/openrewrite/xml/security/DateBoundSuppressionsTest.java:64]: Ah, I see. You can use `String.formatted()` on a multi-line string to do interpolations. Uses the same format specifiers as `String.format()`

File: rewrite-xml/src/main/java/org/openrewrite/xml/security/DateBoundSuppressions.java (hunk 1 above)
Comment (kmccarp) [reply rewrite-xml/src/main/java/org/openrewrite/xml/security/DateBoundSuppressions.java:51]: Added in latest commit, thanks for the suggestion!

File: rewrite-xml/src/main/java/org/openrewrite/xml/security/DateBoundSuppressions.java (hunk 2 above)
Comment (kmccarp) [reply rewrite-xml/src/main/java/org/openrewrite/xml/security/DateBoundSuppressions.java]: Do i just wrap this whole `new Xml.Attribute` in `autoFormat`?

File: rewrite-xml/src/test/java/org/openrewrite/xml/security/DateBoundSuppressionsTest.java (hunk 3 above)
Comment (kmccarp) [reply rewrite-xml/src/test/java/org/openrewrite/xml/security/DateBoundSuppressionsTest.java]: Thanks! Added in latest commit.

File: rewrite-xml/src/main/java/org/openrewrite/xml/security/DateBoundSuppressions.java (hunk 2 above)
Comment (kmccarp) [reply rewrite-xml/src/main/java/org/openrewrite/xml/security/DateBoundSuppressions.java]: Latest commit has the attribute and value both wrapped in autoFormat

File: rewrite-xml/src/main/java/org/openrewrite/xml/security/IsOwaspSuppressionsFile.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        for (Xml.Attribute attribute : root.getAttributes()) {
+            if (attribute.getKeyAsString().equals("xmlns")) {
+                if (attribute.getValueAsString().matches(MATCHER)) {
+                    isOwaspSuppressionFile = true;
+                }
+            }
+        }
+        if (isOwaspSuppressionFile) {
+            return doc.withRoot(doc.getRoot().withMarkers(doc.getRoot().getMarkers().addIfAbsent(new SearchResult(Tree.randomId(), "Found it") {
+                @Override
```
Comment (tkvangorder) [inline rewrite-xml/src/main/java/org/openrewrite/xml/security/IsOwaspSuppressionsFile.java:53]: Curious as to why `equals()` is overridden.

File: rewrite-xml/src/main/java/org/openrewrite/xml/security/DateBoundSuppressions.java [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+        return new DateBoundSuppressionsVisitor();
+    }
+
+    @Override
+    protected TreeVisitor<?, ExecutionContext> getApplicableTest() {
+        return new HasOwaspSuppressionsFile();
+    }
+
+    @Override
+    protected TreeVisitor<?, ExecutionContext> getSingleSourceApplicableTest() {
```
Comment (tkvangorder) [inline rewrite-xml/src/main/java/org/openrewrite/xml/security/DateBoundSuppressions.java:58]: Probably sufficient to just have a single source applicability test

File: rewrite-xml/src/main/java/org/openrewrite/xml/security/RemoveSuppressions.java [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    }
+
+    @Override
+    protected TreeVisitor<?, ExecutionContext> getVisitor() {
+        return new RemoveSuppressionsVisitor();
+    }
+
+    @Override
+    protected TreeVisitor<?, ExecutionContext> getApplicableTest() {
+        return new HasOwaspSuppressionsFile();
```
Comment (tkvangorder) [inline rewrite-xml/src/main/java/org/openrewrite/xml/security/RemoveSuppressions.java]: Same here, looks like you can just get away with a single source applicability test

File: rewrite-xml/src/main/java/org/openrewrite/xml/security/DateBoundSuppressions.java [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+        return new IsOwaspSuppressionsFile();
+    }
+
+    private class DateBoundSuppressionsVisitor extends XmlIsoVisitor<ExecutionContext> {
+        @Override
+        public Xml.Tag visitTag(Xml.Tag tag, ExecutionContext ctx) {
+            Xml.Tag t = super.visitTag(tag, ctx);
+            if (t.getName().equals("suppress")) {
```
Comment (tkvangorder) [inline rewrite-xml/src/main/java/org/openrewrite/xml/security/DateBoundSuppressions.java:66]: probably ok to just use the tag name here, but just be aware there is an XPathMatcher that is a little more explicit when finding a tag in XML. For an example of how these are used see https://github.com/openrewrite/rewrite/blob/main/rewrite-maven/src/main/java/org/openrewrite/maven/MavenVisitor.java#L40-L44

File: rewrite-xml/src/main/java/org/openrewrite/xml/security/DateBoundSuppressions.java [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+                boolean hasUntil = false;
+                List<Xml.Attribute> attributes = t.getAttributes();
+                List<Xml.Attribute> newAttributes = new ArrayList<>();
```
Comment (tkvangorder) [inline rewrite-xml/src/main/java/org/openrewrite/xml/security/DateBoundSuppressions.java:69]: Instead of creating a new list and adding the elements to that new list as you look for the existing attribute, you could instead, just search the existing list for the attribute and if found you can use `ListUtils.concat()` to handle creating the new list and adding the element.

https://github.com/openrewrite/rewrite/blob/main/rewrite-core/src/main/java/org/openrewrite/internal/ListUtils.java#L202-L211

File: rewrite-xml/src/main/java/org/openrewrite/xml/security/RemoveSuppressions.java [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+        return new IsOwaspSuppressionsFile();
+    }
+
+    private static class RemoveSuppressionsVisitor extends XmlIsoVisitor<ExecutionContext> {
+        @Override
+        public Xml.Tag visitTag(Xml.Tag tag, ExecutionContext ctx) {
+            Xml.Tag t = super.visitTag(tag, ctx);
+            if (t.getName().equals("suppressions") && t.getContent() != null) {
+                List<Content> newContent = new ArrayList<>();
+                List<? extends Content> contents = t.getContent();
```
Comment (tkvangorder) [inline rewrite-xml/src/main/java/org/openrewrite/xml/security/RemoveSuppressions.java:64]: Not really a request for a change because I don't think this code will be called a ton and more just informational. If this code was in a hot path, I think you would want to defer the creation of a new list only after you discover you need to remove the suppression.

**Review by tkvangorder — CHANGES_REQUESTED** [review]
Very cool recipe Kevin, just some small suggestion, otherwise looks good!

File: rewrite-xml/src/main/java/org/openrewrite/xml/security/IsOwaspSuppressionsFile.java (hunk 4 above)
Comment (kmccarp) [reply rewrite-xml/src/main/java/org/openrewrite/xml/security/IsOwaspSuppressionsFile.java:53]: by default it checks equality on the id, which would be random in each iteration, which would lead to the marker being added on each cycle.

File: rewrite-xml/src/main/java/org/openrewrite/xml/security/DateBoundSuppressions.java (hunk 5 above)
Comment (kmccarp) [reply rewrite-xml/src/main/java/org/openrewrite/xml/security/DateBoundSuppressions.java:58]: I do now, look at the latest code :)

File: rewrite-xml/src/main/java/org/openrewrite/xml/security/RemoveSuppressions.java (hunk 9 above)
Comment (tkvangorder) [reply rewrite-xml/src/main/java/org/openrewrite/xml/security/RemoveSuppressions.java:64]: Oh, I think you could use `ListUtils.flatMap()` here and if you discover the suppression, you could just return null from the mapping function.

File: rewrite-xml/src/main/java/org/openrewrite/xml/security/DateBoundSuppressions.java (hunk 7 above)
Comment (kmccarp) [reply rewrite-xml/src/main/java/org/openrewrite/xml/security/DateBoundSuppressions.java:66]: the XPathMatcher is a much better idea, will change to that. Thanks!

**Review by tkvangorder — APPROVED** [review]
Looks great!
